dashboard/src/components/Catalog/Catalog.test.tsx

```diff
@@ -196,6 +318,128 @@ describe("filters by application type", () => {
+            isFetching: false,
+            items: [],
+          } as any
+        }
+      />,
+    );
+
+    expect(wrapper.find("CatalogItems").prop("page")).toBe(1);
+    expect(setState).toHaveBeenCalledWith(1);
+    expect(mockUseEffect).toHaveBeenNthCalledWith(3, [false, false]); // [hasRequestedFirstPage, isFetching])
+    expect(fetchCharts).toHaveBeenNthCalledWith(1, "default-cluster", "kubeapps", "", 1, 20, ""); // [hasRequestedFirstPage, isFetching])
+    expect(mockUseEffect).toHaveBeenCalledTimes(7);
+    expect(resetRequestCharts).toHaveBeenNthCalledWith(1);
+  });
+
+  it("sets state params when fetching charts", () => {
+    const fetchCharts = jest.fn();
+    const resetRequestCharts = jest.fn();
+    const setState = jest.fn();
+    const mockUseEffect = jest.fn();
+    spyOnUseState = jest
+      .spyOn(React, "useState")
+      //  @ts-ignore
+      .mockImplementation((init: any) => [init, setState]);
+    spyOnUseEffect = jest
+      .spyOn(React, "useEffect")
+      //  @ts-ignore
+      .mockImplementation((f, n) => f(mockUseEffect(n)));
+
+    const wrapper = mountWrapper(
+      defaultStore,
+      <Catalog
+        {...populatedProps}
+        fetchCharts={fetchCharts}
+        resetRequestCharts={resetRequestCharts}
+        charts={
+          {
+            ...defaultChartState,
+            hasFinishedFetching: false,
+            isFetching: true, // true
+            items: [chartItem, chartItem2],
+          } as any
+        }
+      />,
+    );
+    expect(wrapper.find("CatalogItems").prop("page")).toBe(1);
+    expect(setState).toHaveBeenCalledWith(1);
+    expect(fetchCharts).toHaveBeenCalledWith("default-cluster", "kubeapps", "", 1, 20, "");
+    expect(mockUseEffect).toHaveBeenCalledTimes(6);
+    expect(resetRequestCharts).toHaveBeenCalledWith();
+  });
+
+  it("sets state params when charts have been fetched", () => {
+    const fetchCharts = jest.fn();
+    const resetRequestCharts = jest.fn();
+    const setState = jest.fn();
+    const mockUseEffect = jest.fn();
+    spyOnUseState = jest
+      .spyOn(React, "useState")
+      //  @ts-ignore
+      .mockImplementation((init: any) => [init, setState]);
+    spyOnUseEffect = jest
+      .spyOn(React, "useEffect")
+      //  @ts-ignore
+      .mockImplementation((f, n) => f(mockUseEffect(n)));
+
+    const wrapper = mountWrapper(
+      defaultStore,
+      <Catalog
+        {...populatedProps}
+        fetchCharts={fetchCharts}
+        resetRequestCharts={resetRequestCharts}
+        charts={
+          {
+            ...defaultChartState,
+            hasFinishedFetching: true,
+            isFetching: false,
+            items: [chartItem, chartItem2],
+          } as any
+        }
+      />,
+    );
+    expect(wrapper.find("CatalogItems").prop("page")).toBe(1);
+    expect(setState).toHaveBeenCalledWith(1);
+    expect(mockUseEffect).toHaveBeenCalledTimes(9);
+    expect(resetRequestCharts).toHaveBeenNthCalledWith(1);
+  });
+});
+
```

Comment:
I wanted to check if it's possible to mock the observer to test that at least we are changing the page. I managed to do so:

```js
  it("changes page", () => {
    const setState = jest.fn();
    const setPage = jest.fn();
    const charts = {
      ...defaultChartState,
      hasFinishedFetching: false,
      isFetching: false,
      items: [],
    } as any;
    spyOnUseState = jest
      .spyOn(React, "useState")
      //  @ts-ignore
      .mockImplementation((init: any) => {
        if (init === false) {
          // Mocking the result of hasLoadedFirstPage to simulate that is already loaded
          return [true, setState];
        }
        if (init === 1) {
          // Mocking the result of setPage to ensure it's called
          return [1, setPage];
        }
        return [init, setState];
      });

    // Mock intersection observer
    const observe = jest.fn();
    const unobserve = jest.fn();
    window.IntersectionObserver = jest.fn(callback => {
      (callback as (e: any) => void)([{ isIntersecting: true }]);
      return { observe, unobserve } as any;
    });

    mountWrapper(defaultStore, <Catalog {...populatedProps} charts={charts} />);
    expect(setPage).toHaveBeenCalledWith(2);
  });
```

Ideally we should check that `fetchCharts` is called with the second page, rather than mocking `setPage` but if we don't mock `setPage`, it enters an infinite loop increasing the page.

Reply:
BTW, you may need to also mock `window.IntersectionObserverEntry`, I commented that part of the code when trying this

Reply:
FTR:
The condition was: `if ("IntersectionObserver" in window && "IntersectionObserverEntry" in window && "isIntersecting" in window.IntersectionObserverEntry.prototype)`

Since mocking `window.IntersectionObserverEntry.prototype` is causing problems, we decided to get rid of this condition.
In the future, if we decide to implement a fallback mechanism (fetch all pages instead of one-by-one), we can add it back.